src/Maestro/ReleasePipelineRunner/ReleasePipelineRunner.cs

```diff
@@ -81,12 +81,52 @@ public async Task<TimeSpan> RunAsync(CancellationToken cancellationToken)
                     if (maybeItem.HasValue)
                     {
                         ReleasePipelineRunnerItem item = maybeItem.Value;
-                        using (Logger.BeginScope(
-                            $"Triggering release pipelines associated with channel {item.ChannelId} for build {item.BuildId}.",
-                            item.BuildId,
-                            item.ChannelId))
+
+                        Build build = await Context.Builds
+                            .Where(b => b.Id == item.BuildId).FirstOrDefaultAsync();
+
+                        if (build == null)
+                        {
+                            Logger.LogError($"Could not find the specified BAR Build {item.BuildId} to run a release pipeline.");
+                        }
+                        else if (build.AzureDevOpsBuildId == null)
                         {
-                            await RunAssociatedReleasePipelinesAsync(item.BuildId, item.ChannelId, cancellationToken);
+                            // If something uses the old API version we won't have this information available.
+                            // This will also be the case if something adds an existing build (created using
+                            // the old API version) to a channel
+                            Logger.LogInformation($"barBuildInfo.AzureDevOpsBuildId is null for BAR Build.Id {build.Id}.");
+                        }
+                        else
+                        {
+                            AzureDevOpsClient azdoClient = await GetAzureDevOpsClientForAccount(build.AzureDevOpsAccount);
+
+                            var azdoBuild = await azdoClient.GetBuildAsync(
+                                build.AzureDevOpsAccount,
+                                build.AzureDevOpsProject,
+                                build.AzureDevOpsBuildId.Value);
+
+                            if (azdoBuild.Status.Equals("completed", StringComparison.OrdinalIgnoreCase))
+                            {
+                                if (azdoBuild.Status.Equals("succeeded", StringComparison.OrdinalIgnoreCase))
+                                {
+                                    using (Logger.BeginScope(
+                                        $"Triggering release pipelines associated with channel {item.ChannelId} for build {item.BuildId}.",
+                                        item.BuildId,
+                                        item.ChannelId))
+                                    {
+                                        await RunAssociatedReleasePipelinesAsync(item.BuildId, item.ChannelId, cancellationToken);
+                                    }
+                                }
+                                else
+                                {
+                                    Logger.LogError($"Tried to trigger release pipeline for a non-succeeded build: {item.BuildId}");
```

Comment:
Does this retry once the build succeeds?

Reply:
Non-succeeded here means the build has finished execution but the end result wasn't "success". We don't need to retry processing builds in this state.

Reply:
Hmm, I think it's an interesting scenario:

* Build attempt 1 succeeds in publishing to BAR, but fails in a subsequent job.
* Dev goes ahead and retries from AzDO.
* Attempt 2 won't run the publish to BAR job again, since retry only runs the failed jobs from the 1st attempt.

Since we simply skipped over that build, we'll never actually try again. Then the user is just confused because their build was green, but no publishing actually happened.

Reply:
That's indeed a complicated scenario.. retries can happen anytime after the build has finished. How long we'll keep checking if the build succeeded?

Reply:
The ideal way I envision this is:

We dequeue an item:
* If build is in finished successful state, process it
* Else put item in the back of the queue again, possibly with an exponential delay in when the queue message is visible (do this for what, 10 attempts or so?)

That way we'll keep trying unfinished builds for a certain number of attempts, and since the retry is exponential, enough time will have passed that if we did all the attempts, the build is probably obsolete by then.

Note that this is all hand waving, I don't know if we have enough control over the queue to do that.

Reply:
I think this approach or something close to that (due to the limitations in the queue API) should be sufficient because: 1) usually we retry builds that finished "recently" (few hours) ago at max?! 2) This will be temporary.